File: php/BinaryTree/BTree.php
<?php
/**
 * @date 2020/5/5
 */

class BTree
{
    public $val = null;
    public $left = null;
    public $right = null;
    function __construct($value)
    {
        $this->val = $value;
    }
}

class Creator
{
    public $arr = [];

    public static function preOrder($root)
    {
        if (!empty($root)) {
            $func = __FUNCTION__;
            echo $root->val . '';
            $func($root->left);
            $func($root->right);
        }
    }

    public function inorder($root)
    {
        if (!empty($root)) {
            $func = __FUNCTION__;
            $func($root->left);
            echo $root->val . '';
            $func($root->right);
        }
    }

    public static function postOrder($root)
    {
        if (!empty($root)) {
            $func = __FUNCTION__;

            $func($root->left);
            $func($root->right);
            echo $root->val . '';
        }
    }

    public static function levelOrder($root)
    {
        self::order($root, 0);
        return self::$arr;
    }

    public function order($root, $level)
    {
        if (empty($root)) {
            return $root;
        }

        $this->arr[$level][] = $root->val;
        $level++;

        if (!empty($root->left)) {
            $this->order($root->left, $level);
        }

        if (!empty($root->right)) {
            $this->order($root->right, $level);
        }
    }

    /**
     * 层序遍历非递归
     * @param $root
     */
    public static function levelOrder1($root)
    {
        if (empty($root)) {
            return [];
        }

        $res = [];
        $queue = [];

        array_push($queue, $root);
        $level = 0;
        while ($count = count($queue)) {
            for ($i = 0; $i < $count; $i++) {
                $node = array_shift($queue);
                $res[$level][] = $node->val;

                if ($node->left) {
                    $queue[] = $node->left;
                }

                if ($node->right) {
                    $queue[] = $node->right;
                }
            }
            $level++;
        }

        return $res;
    }

    /** 先序遍历非递归 */
    public static function preOrder1($root)
    {
        if (empty($root)) {
            return [];
        }

        $res = [];
        $stack = [];
        array_push($stack, $root);

        while (count($stack)) {
            $node = array_pop($stack);
            $res[] = $node->val;
            if (!empty($node->right)) {
                array_push($stack, $node->right);
            }
            if (!empty($node->left)) {
                array_push($stack, $node->left);
            }
        }

        return $res;
    }

    /**
     * 中序遍历 迭代
     * @param $root
     *
     * @return array
     */
    public static function inorderTraversal($root)
    {
        if (empty($root)) {
            return [];
        }

        $res = [];
        $stack = [];
        while (!empty($root) || !empty($stack)) {
            if (!empty($root)) {
                array_push($stack, $root);
                $root = $root->left;
            } else {
                $root = array_pop($stack);
                $res[] = $root->val;
                $root = $root->right;
            }
        }

        return $res;
    }
}

File: BinaryTree/BTree.go
package BinaryTree

import "fmt"

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func CreateNode(item int) *TreeNode {

	if item == -1 {
		return nil
	}

	return &TreeNode{item, nil, nil}
}

//层序创建二叉树
func CreateByBreadthFirstSearch(inputList []int) *TreeNode {

	queue := []*TreeNode{}
	root := CreateNode(inputList[0])
	queue = append(queue, root)

	inputList = inputList[1:]

	for len(inputList) > 0 {
		curNode := queue[0]
		queue = queue[1:]

		curNode.Left = CreateNode(inputList[0])
		if curNode.Left != nil {
			queue = append(queue, curNode.Left)
		}
		inputList = inputList[1:]

		curNode.Right = CreateNode(inputList[0])
		if curNode.Right != nil {
			queue = append(queue, curNode.Right)
		}
		inputList = inputList[1:]

	}

	return root
}

func Inorder(root *TreeNode) {
	if root == nil {
		return
	}

	Inorder(root.Left)
	fmt.Printf("%v\n", root.Val)
	Inorder(root.Right)
}

//前序遍历创建二叉树
func CreateBTreeByPreOrder(input []int)  *TreeNode {
	if input == nil {
		return nil
	}

	var pre func(input []int, i int) *TreeNode
	pre = func(input []int, i int) *TreeNode {
		if i >= len(input) {
			return nil
		}

		root := CreateNode(input[i])
		i++
		root.Left = pre(input, i)
		root.Right = pre(input, i)
		return root
	}

	root := pre(input, 0)
	return root
}

/**
 * 二叉树层序遍历
 */
func levelOrder(root *TreeNode) (res [][]int) {

	queue := []*TreeNode{
		root,
	}

	for queue != nil {
		tmp := queue
		queue = nil

		level := []int{}
		for _, node := range tmp {
			level = append(level, node.Val)

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		res = append(res, level)
	}

	return
}

